internal/remote/modulereleasemeta_syncer.go

```diff
@@ -0,0 +1,114 @@
+package remote
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kyma-project/lifecycle-manager/api/v1beta2"
+	"github.com/kyma-project/lifecycle-manager/internal/util/collections"
+	"github.com/kyma-project/lifecycle-manager/pkg/util"
+)
+
+// moduleReleaseMetaSyncWorker is an interface for worker synchronizing ModuleReleaseMetas from KCP to SKR.
+type moduleReleaseMetaSyncWorker interface {
+	SyncConcurrently(ctx context.Context, kcpModules []v1beta2.ModuleReleaseMeta) error
+	DeleteConcurrently(ctx context.Context, runtimeModules []v1beta2.ModuleReleaseMeta) error
+}
+
+// moduleReleaseMetaSyncWorkerFactory is a factory function for creating new moduleReleaseMetaSyncWorker instance.
+type moduleReleaseMetaSyncWorkerFactory func(kcpClient, skrClient client.Client, settings *Settings) moduleReleaseMetaSyncWorker
+
+// moduleReleaseMetaSyncer provides a top-level API for synchronizing ModuleReleaseMetas from KCP to SKR.
+// It expects a ready-to-use client to the KCP and SKR cluster.
+type moduleReleaseMetaSyncer struct {
+	kcpClient           client.Client
+	skrClient           client.Client
+	settings            *Settings
+	syncWorkerFactoryFn moduleReleaseMetaSyncWorkerFactory
+}
+
+func newModuleReleaseMetaSyncer(kcpClient, skrClient client.Client, settings *Settings) *moduleReleaseMetaSyncer {
+	var syncWokerFactoryFn moduleReleaseMetaSyncWorkerFactory = func(kcpClient, skrClient client.Client, settings *Settings) moduleReleaseMetaSyncWorker {
+		return newModuleReleaseMetaConcurrentWorker(kcpClient, skrClient, settings)
+	}
+
+	return &moduleReleaseMetaSyncer{
+		kcpClient:           kcpClient,
+		skrClient:           skrClient,
+		settings:            settings,
+		syncWorkerFactoryFn: syncWokerFactoryFn,
+	}
+}
+
+// SyncToSKR first lists all currently available ModuleReleaseMetas in the Runtime.
+// If there is a NoMatchError, it will attempt to install the CRD but only if there are available crs to copy.
+// It will use a 2 stage process:
+// 1. All ModuleReleaseMeta that have to be created based on the ModuleReleaseMetas existing in the Control Plane.
+// 2. All ModuleReleaseMeta that have to be removed as they are not existing in the Control Plane.
+// It uses Server-Side-Apply Patches to optimize the turnaround required.
+func (mts *moduleReleaseMetaSyncer) SyncToSKR(ctx context.Context, kyma types.NamespacedName, kcpModuleReleases []v1beta2.ModuleReleaseMeta) error {
```

Comment:
The kyma parameter is not used here

Reply:
Good catch! Leftover from intermediate refactoring steps.